package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+fn do_build_minimal(config: &Config) -> Result<()> {
+    let server = &config.servers[&config.builder.server];
+    let cmd = format!(
+        "cd {} && git checkout {} && cargo build -p {} -p {}",
+        config.builder.omicron_path,
+        config.builder.git_treeish,
+        "omicron-package",
+        "omicron-sled-agent"
+    );
+    ssh_exec(&server, &cmd, false)
+}
+
+fn do_package(
+    config: &Config,
+    artifact_dir: PathBuf,
+    release: bool,
+) -> Result<()> {
+    let server = &config.servers[&config.builder.server];
+    let mut cmd = String::new();
+    let mut release_flag = "";
+    if release {
+        release_flag = "--release";
+    }
+    let cmd_path = "./target/debug/omicron-package";
+    let artifact_dir = artifact_dir
+        .to_str()
+        .ok_or_else(|| FlingError::BadString("artifact_dir".to_string()))?;
+
+    // We use a bash login shell to get a proper environment, so we have a path to
+    // postgres, and $DEP_PQ_LIBDIRS is filled in. This is required for building
+    // nexus.
+    //
+    // See https://github.com/oxidecomputer/omicron/blob/8757ec542ea4ffbadd6f26094ed4ba357715d70d/rpaths/src/lib.rs
+    write!(
+        &mut cmd,
+        "bash -lc 'cd {} && git checkout {} && {} package --out {} {}'",
+        config.builder.omicron_path,
+        config.builder.git_treeish,
+        cmd_path,
+        &artifact_dir,
+        release_flag
+    )?;
+
+    ssh_exec(&server, &cmd, false)
+}
+
+fn do_check(config: &Config) -> Result<()> {
+    let server = &config.servers[&config.builder.server];
+    let mut cmd = String::new();
+    let cmd_path = "./target/debug/omicron-package";
+
+    write!(
+        &mut cmd,
+        "bash -lc 'cd {} && git checkout {} && {} check'",
+        config.builder.omicron_path, config.builder.git_treeish, cmd_path,
+    )?;
+
+    ssh_exec(&server, &cmd, false)
+}
+
+fn do_uninstall(
+    config: &Config,
+    artifact_dir: PathBuf,
+    install_dir: PathBuf,
+) -> Result<()> {
+    let mut deployment_src = PathBuf::from(&config.deployment.staging_dir);
+    deployment_src.push(&artifact_dir);
+    for server_name in &config.deployment.servers {
+        let server = &config.servers[server_name];
+        // Run `omicron-package uninstall` on the deployment server
+        let cmd = format!(
+            "cd ~/{} && pfexec ./omicron-package uninstall --in ~/{} --out {}",
+            config.deployment.staging_dir,
+            deployment_src.to_string_lossy(),
+            install_dir.to_string_lossy()
+        );
+        println!("$ {}", cmd);
+        ssh_exec(&server, &cmd, true)?;
+    }
+    Ok(())
+}
+
+fn do_install(config: &Config, artifact_dir: &Path, install_dir: &Path) {
+    let builder = &config.servers[&config.builder.server];
+    let mut pkg_dir = PathBuf::from(&config.builder.omicron_path);
+    pkg_dir.push(artifact_dir);
+    let pkg_dir = pkg_dir.to_string_lossy();
+    let pkg_dir = &pkg_dir;
+
+    thread::scope(|s| {
```

Comment:
More curious than anything, but why does this need to be a scoped thread?

Reply:
I'm not sure it's **needed**, but it allows borrowing all the parameters in multiple threads without cloning or using Arcs. Also, when I first wrote it, I didn't bother joining handles as they can't outlive the scope. I only added that join code later to provide status about completed tasks.